src/Http/Http/src/Timeouts/RequestTimeoutsMiddleware.cs

```diff
@@ -0,0 +1,135 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+
+namespace Microsoft.AspNetCore.Http.Timeouts;
+
+internal sealed class RequestTimeoutsMiddleware
+{
+    private readonly RequestDelegate _next;
+    private readonly ICancellationTokenLinker _cancellationTokenProvider;
+    private readonly ILogger<RequestTimeoutsMiddleware> _logger;
+    private readonly IOptionsMonitor<RequestTimeoutOptions> _options;
+
+    public RequestTimeoutsMiddleware(
+        RequestDelegate next,
+        ICancellationTokenLinker cancellationTokenProvider,
+        ILogger<RequestTimeoutsMiddleware> logger,
+        IOptionsMonitor<RequestTimeoutOptions> options)
+    {
+        _next = next;
+        _cancellationTokenProvider = cancellationTokenProvider;
+        _logger = logger;
+        _options = options;
+    }
+
+    public Task Invoke(HttpContext context)
+    {
+        if (Debugger.IsAttached)
+        {
+            return _next(context);
+        }
+
+        if (context.RequestAborted.IsCancellationRequested)
```

Comment:
What do you think about this? Skipping the middleware if the original cancellation token is cancelled already.

Reply:
That shouldn't come up much, but it's ok to have the check. Check just before SetTimeoutAsync so we can avoid lazy allocations of RequestAborted if not needed.